### PR TITLE
feat: create with context

### PR DESCRIPTION
**Pull request by mfw78**
This PR:

1. Allows for setting an on-chain context at the time of mining for the merkle root set / single order creation transaction.

### CODE REVIEW

File: src/interfaces/IConditionalOrder.sol [hunk 1]
```diff
@@ -30,6 +31,7 @@ interface IConditionalOrder {
      * @param sender the `msg.sender` of the transaction
      * @param _hash the hash of the order
      * @param domainSeparator the domain separator used to sign the order
+     * @param ctx the context of the order (bytes32(0) if a merkle tree is used, otherwise H(params))
```
Comment (fleupold) [inline src/interfaces/IConditionalOrder.sol:34]: Maybe specify the intent here that this value can be used to look up order specific storage in composable cow?

**Review by fleupold — DISMISSED** [review]
Looks good to me. Are you also going to check in a `CurrrentBlockTimestampFactory`?

File: src/interfaces/IConditionalOrder.sol (hunk 1 above)
Comment (mfw78) [reply src/interfaces/IConditionalOrder.sol:34]: Done

File: test/ComposableCoW.twap.t.sol [hunk 2]
```diff
@@ -225,6 +231,40 @@ contract ComposableCoWTwapTest is BaseComposableCoWTest {
         twap.getTradeableOrder(address(0), address(0), bytes32(0), abi.encode(o), bytes(""));
     }
 
+    function test_getTradeableOrder_e2e_fuzz_WithContext(uint32 _blockTimestamp, uint256 currentTime) public {
+        // guard against overflows
+        vm.assume(_blockTimestamp < type(uint32).max);
+        vm.assume(currentTime < type(uint32).max);
+        // guard against revert before start
+        vm.assume(_blockTimestamp < currentTime);
```
Comment (fleupold) [inline test/ComposableCoW.twap.t.sol:239]: Not necessarily as a fuzz test, but might be nice to have one test case showing that blockTimestamp is actually used a start time (technically startTime 0 would have worked in the old codebase wouldn't it?)

**Review by fleupold — APPROVED** [review]
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2FHw8vYF4DNRCKY%2Fgiphy.gif%3Fcid%3D2972a2b15d1d81cf6d6d7963363a4f7b%26rid%3Dgiphy.gif&token=shipit" data-gitmeme-token="shipit"></a>

File: test/ComposableCoW.twap.t.sol (hunk 2 above)
Comment (mfw78) [reply test/ComposableCoW.twap.t.sol:239]: Absolute fair - incoming!

File: src/ComposableCoW.sol [hunk 3]
```diff
@@ -91,6 +113,28 @@ contract ComposableCoW is ISafeSignatureVerifier {
...
+     * @param data Implementation specific off-chain data
+     * @param dispatch Whether to dispatch the `ConditionalOrderCreated` event
+     */
+    function createWithContext(
+        IConditionalOrder.ConditionalOrderParams calldata params,
+        IValueFactory factory,
+        bytes calldata data,
+        bool dispatch
+    ) external {
+        create(params, dispatch);
```
Comment (anxolin) [inline src/ComposableCoW.sol:129]: not sure if there's gas optimizations possible here, or if it's not even worth caring, but I just noted that we hash the params twice (once inside the good old `create` and the other to store in the `cabinet`

File: src/ComposableCoW.sol [hunk 4]
```diff
@@ -130,19 +174,26 @@ contract ComposableCoW is ISafeSignatureVerifier {
         PayloadStruct memory _payload = abi.decode(payload, (PayloadStruct));
 
         // Check if the order is authorised
-        _auth(address(safe), _payload.params, _payload.proof);
+        bytes32 ctx = _auth(address(safe), _payload.params, _payload.proof);
```
Comment (anxolin) [inline src/ComposableCoW.sol:177]: I didn't fully get what is `ctx` for. From what i can tell, is returning the hash that identifies the order, and only for `singleOrders`. But this is used later for `getTradeableOrder`. isn't it relevant for the Merkle root flow?
I guess I'm missing something

File: src/ComposableCoW.sol [hunk 5]
```diff
@@ -229,34 +280,43 @@ contract ComposableCoW is ISafeSignatureVerifier {
...
 
-        if (!(proof.length != 0 || singleOrders[owner][hash(params)])) {
-            revert SingleOrderNotAuthed();
+            // Check if the proof is valid
+            if (!MerkleProof.verify(proof, roots[owner], leaf)) {
+                revert ProofNotAuthed();
+            }
+        } else {
+            // Check if the order is authorised
+            ctx = hash(params);
```
Comment (anxolin) [inline src/ComposableCoW.sol:295]: just double-ckecing if this should be only in the else, or should be done outside of the IF

File: src/types/twap/TWAP.sol [hunk 6]
```diff
@@ -30,7 +38,16 @@ contract TWAP is BaseConditionalOrder {
...
          * settled once.
          */
-        order = TWAPOrder.orderFor(abi.decode(staticInput, (TWAPOrder.Data)));
+        TWAPOrder.Data memory twap = abi.decode(staticInput, (TWAPOrder.Data));
+
+        /**
+         * @dev If `twap.t0` is set to 0, then get the start time from the context.
+         */
+        if (twap.t0 == 0) {
+            twap.t0 = uint256(composableCow.cabinet(owner, ctx));
```
Comment (anxolin) [inline src/types/twap/TWAP.sol:47]: ❤️

**Review by anxolin — COMMENTED** [review]
👍 Great changes!

File: src/ComposableCoW.sol (hunk 5 above)
Comment (mfw78) [reply src/ComposableCoW.sol:295]: Yep, it should only be done in the `else`. This takes advantage of the `ctx` return parameter already being initialised to the `bytes32(0)` (zero) value, which is the case to return when considering a Merkle flow.

File: src/BaseConditionalOrder.sol [hunk 7]
```diff
@@ -20,11 +20,12 @@ abstract contract BaseConditionalOrder is IConditionalOrderGenerator {
         address sender,
         bytes32 _hash,
         bytes32 domainSeparator,
+        bytes32 ctx,
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/BaseConditionalOrder.sol:23]: Sure. Given the `@inheritdoc IConditionalOrder`, this picks up the natdoc from `IConditionalOrder.verify` which contains the `ctx` and other parameter documentation. :+1:

**Pull request merged**